src/eos-updater-poll-lan.c

```diff
@@ -0,0 +1,920 @@
+              latest_timestamp = 0;
+              g_ptr_array_set_size (swbfs_with_latest_commit, 0);
+              g_ptr_array_set_size (urls, 0);
+              g_clear_object (&latest_extensions);
+            }
+        }
+      if (latest_checksum == NULL)
+        {
+          latest_checksum = g_steal_pointer (&checksum);
+          latest_commit = g_steal_pointer (&commit);
+          latest_timestamp = timestamp;
+          g_ptr_array_add (swbfs_with_latest_commit, g_object_ref (swbf));
+          g_ptr_array_add (urls, g_steal_pointer (&url_override));
+          latest_extensions = g_steal_pointer (&extensions);
+        }
+    }
+
+  if (latest_checksum != NULL)
+    {
+      g_set_object (&latest_extensions->branch_file, newest);
+      *out_info = eos_update_info_new (latest_checksum,
+                                       latest_commit,
+                                       refspec,
+                                       original_refspec,
+                                       (gchar **)g_ptr_array_free (g_steal_pointer (&urls),
+                                                                   FALSE),
+                                       latest_extensions);
+    }
+  *out_metrics = g_steal_pointer (&metrics);
+
+  return TRUE;
+}
+
+static gboolean
+check_lan_updates (LanData *lan_data,
+                   GPtrArray *found_services)
+{
+  g_autoptr(GPtrArray) valid_services = NULL;
+  g_autoptr(EosBranchFile) branch_file = NULL;
+
+  if (!filter_services (lan_data,
+                        found_services,
+                        &valid_services,
+                        &lan_data->error))
+    {
+      message ("Failed to filter services: %s",
+               lan_data->error->message);
+      return TRUE;
+    }
+
+  if (valid_services->len == 0)
+    {
+      message ("No valid LAN servers found");
+      return TRUE;
+    }
+
+  branch_file = get_newest_branch_file (lan_data, valid_services);
+  if (branch_file == NULL)
+    {
+      message ("No valid branch file found");
+      return TRUE;
+    }
+
+  if (!get_update_info_from_swbfs (lan_data,
+                                   valid_services,
+                                   branch_file,
+                                   &lan_data->info,
+                                   &lan_data->metrics,
+                                   &lan_data->error))
+    {
+      message ("Failed to get the latest update info: %s",
+               lan_data->error->message);
+      return TRUE;
+    }
+
+  return TRUE;
+}
+
+static void
+discoverer_callback (EosAvahiDiscoverer *discoverer,
+                     GPtrArray *found_services,
+                     gpointer lan_data_ptr,
+                     GError *error)
+{
+  LanData *lan_data = lan_data_ptr;
+  gboolean quit_loop = TRUE;
+
+  lan_data->error = g_steal_pointer (&error);
+  if (lan_data->error == NULL)
+    quit_loop = check_lan_updates (lan_data, found_services);
+
```

Comment:
`else` would be wrong here. `check_lan_updates` can set the `lan_data->error` member too.